### PR TITLE
Fixed path encoding in AgaviXmlConfigParser::xinclude()

### PR DESCRIPTION
**Pull request by thomasbachem**
Fixes the whitespace issues with XIncludes outlined by [Agavi WTF](/agavi/agavi/wiki/WTF)

### CODE REVIEW

File: src/config/AgaviXmlConfigParser.class.php [hunk 1]
```diff
@@ -494,21 +494,25 @@ public static function executeCompilation(AgaviXmlConfigDomDocument $document, $
...
 					$glob = glob($parts[0], GLOB_BRACE | GLOB_NOSORT);
 					if($glob) {
 						$glob = array_unique($glob); // it could be that someone used /path/to/{Foo,*}/burp.xml so Foo would come before all others, that's why we need to remove duplicates as the * would match Foo again
+						sort($glob);
 						foreach($glob as $path) {
 							$new = $element->cloneNode(true);
-							$new->setAttribute('href', $path . (isset($parts[1]) ? '#' . $parts[1] : ''));
+							$new->setAttribute('href', rawurlencode($path) . (isset($parts[1]) ? '#' . $parts[1] : ''));
 							$element->parentNode->insertBefore($new, $element);
+							++$i;
```
Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: What's the point of this? That code needs some comments because it's totally unclear why you're apparently skipping an element on the outer loop. Only you know, right now, but won't be able to say in a few months.

File: src/config/AgaviXmlConfigParser.class.php [hunk 2]
```diff
@@ -494,21 +494,25 @@ public static function executeCompilation(AgaviXmlConfigDomDocument $document, $
...
 				$attribute = $element->getAttributeNode('href');
 				$parts = explode('#', $attribute->nodeValue, 2);
 				$parts[0] = str_replace('\\', '/', AgaviToolkit::expandDirectives($parts[0]));
-				$attribute->nodeValue = implode('#', $parts);
+				$attribute->nodeValue = rawurlencode($parts[0]) . (isset($parts[1]) ? '#' . $parts[1] : '');
 				if(strpos($parts[0], '*') !== false || strpos($parts[0], '{') !== false) {
 					$glob = glob($parts[0], GLOB_BRACE | GLOB_NOSORT);
 					if($glob) {
 						$glob = array_unique($glob); // it could be that someone used /path/to/{Foo,*}/burp.xml so Foo would come before all others, that's why we need to remove duplicates as the * would match Foo again
+						sort($glob);
```
Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: What's this for? It seems like an unrelated change?

File: src/config/AgaviXmlConfigParser.class.php [hunk 3]
```diff
@@ -494,21 +494,25 @@ public static function executeCompilation(AgaviXmlConfigDomDocument $document, $
 	 */
 	public static function xinclude(AgaviXmlConfigDomDocument $document)
 	{
-		// replace %lala% directives in XInclude href attributes
-		foreach($document->getElementsByTagNameNS(self::NAMESPACE_XINCLUDE_2001, 'include') as $element) {
+		$elements = $document->getElementsByTagNameNS(self::NAMESPACE_XINCLUDE_2001, 'include');
+		$length = $elements->length;
+		for($i = 0; $i < $length; $i++) {
```
Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: Why the for() loop? Could you not `foreach($elements as &element)`?

File: src/config/AgaviXmlConfigParser.class.php (hunk 1 above)
Comment (thomasbachem) [inline src/config/AgaviXmlConfigParser.class.php]: True, I'm just not used to see comments in Agavi code :p.

It was actually buggy before, because the `DOMNodeList` returned by `getElementsByTagNameNS()` has some auto-updating magic (see http://php.net/manual/en/class.domnodelist.php#83178). So we need to use a `for` loop and the `++$i` to skip the newly added elements.

File: src/config/AgaviXmlConfigParser.class.php (hunk 2 above)
Comment (thomasbachem) [inline src/config/AgaviXmlConfigParser.class.php]: Oups :)

File: src/config/AgaviXmlConfigParser.class.php (hunk 3 above)
Comment (thomasbachem) [inline src/config/AgaviXmlConfigParser.class.php]: See below.

File: src/config/AgaviXmlConfigParser.class.php (hunk 1 above)
Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: Can you please open two separate PRs then, one for the looping fix with a test case, and one for this path encoding change?

Comment (thomasbachem) [inline src/config/AgaviXmlConfigParser.class.php]: A test case makes no sense as the double looping didn't cause any side effects previous to the escaping, it was just not optimal.

Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: Two separate commits then at least?

Comment (thomasbachem) [inline src/config/AgaviXmlConfigParser.class.php]: Ok.

File: src/config/AgaviXmlConfigParser.class.php (hunk 3 above)
Comment (dzuelke) [inline src/config/AgaviXmlConfigParser.class.php]: Then add a comment above the `for()`...

```
// we can't foreach() over the DomNodeList as we're modifying it further below
// see http://php.net/manual/en/class.domnodelist.php#83178
```
